File: LastSemesterInfo/AYDA Laboratorio/AYDA_Examen_4_geometria_computacional_(A2026)/linux_32/cpp/template.cpp
/**
 * @file template.cpp
 * @brief Examen Práctico 4: Emergencia en La Hechicera (Geometría Computacional)
 * Facultad de Ingeniería - ULA, Mérida.
 */

#include <vector>
#include <stdexcept>
#include <cmath>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <vector>
#include <list>
#include <tuple>
#include "Definitions.hpp"
#include "Geom.hpp"

using namespace std;
using namespace geom_lib;

// ========================================================================
// EJERCICIO 1: Línea de Contención
// ========================================================================
/**
 * @brief Construye la frontera óptima alrededor del fuego y verifica que las
 * instalaciones de la ULA estén a salvo.
 * @throws std::runtime_error("LA_HECHICERA_COMPROMETIDA") si una estación está en peligro.
 * * HINT: Recuerda que la frontera más eficiente para encerrar un conjunto de puntos
 * sin concavidades redundantes tiene un nombre específico en geometría.
 */
Polygon build_secure_perimeter(const vector<Point2D>& fire_points, const vector<Point2D>& stations) {
    // TODO: Implementar la construcción del perímetro y la validación de seguridad.
    throw std::logic_error("NOT_IMPLEMENTED");
}

// ========================================================================
// EJERCICIO 2: Huella de Destrucción por Viento
// ========================================================================
/**
 * @brief Calcula el área de la superficie total amenazada considerando el viento.
 * * HINT: Un polígono no es más que una secuencia ordenada de vértices. Si conoces
 * las coordenadas de los vértices del nuevo macro-polígono, existe un teorema
 * algebraico para hallar su área exacta sin importar los polígonos que lo formen.
 */
double calculate_buffer_zone_area(const Polygon& p_fire, const Vector2D& wind_vector) {
    // TODO: Implementar el cálculo de la huella combinada y su área.
    throw std::logic_error("NOT_IMPLEMENTED");
}

// ========================================================================
// EJERCICIO 3: Escuadrón de Drones Estáticos
// ========================================================================
/**
 * @brief Encuentra la combinación de orientaciones de drones que maximiza
 * la vigilancia de focos únicos de incendio.
 * * HINT: Dado que el número de drones es pequeño (N <= 8) y el espacio de
 * decisiones es discreto (4 opciones por dron), evalúa cómo estructurar
 * tu algoritmo para explorar el árbol de posibilidades sin omitir ninguna.
 */
vector<Vector2D> optimize_camera_angles(
    const vector<Point2D>& fire_points,
    const vector<Point2D>& camera_positions,
    double radius,
    double alpha
) {
    // TODO: Implementar la búsqueda de la configuración óptima.
    throw std::logic_error("NOT_IMPLEMENTED");
}


// ========================================================================
// I/O BOILERPLATE (NO MODIFICAR ESTA SECCIÓN)
// Canal de comunicación automatizado con el Grader.
// ========================================================================

int main(int argc, char* argv[]) {

    if (argc < 5) return 1;

    string exercise = argv[1];
    ofstream sol(argv[4]);

    try {
        if (exercise == "ex1") {
            ifstream infile(argv[2]);
            int num_fires; infile >> num_fires;
            vector<Point2D> fire_points;
            for (int i = 0; i < num_fires; ++i) {
                double x, y; infile >> x >> y;
                fire_points.push_back(Point2D(x, y));
            }
            int num_stations; infile >> num_stations;
            vector<Point2D> stations;
            for (int i = 0; i < num_stations; ++i) {
                double x, y; infile >> x >> y;
                stations.push_back(Point2D(x, y));
            }
            infile.close();

            Polygon perimeter = build_secure_perimeter(fire_points, stations);
            perimeter.for_each_vertex([&](const Point2D& v) {
                sol << fixed << setprecision(4) << v.get_x() << " " << v.get_y() << " ";
            });
            sol << "\n";
        }
        else if (exercise == "ex2") {
            ifstream infile(argv[2]);
            int num_vertices; infile >> num_vertices;
            Polygon p_fire;
            for (int i = 0; i < num_vertices; ++i) {
                double x, y; infile >> x >> y;
                p_fire.add_vertex(Point2D(x, y));
            }
            infile.close();

            ifstream params(argv[3]);
            double wx, wy; params >> wx >> wy;
            params.close();

            double area = calculate_buffer_zone_area(p_fire, Vector2D(wx, wy));
            sol << fixed << setprecision(6) << area << "\n";
        }
        else if (exercise == "ex3") {
            ifstream infile(argv[2]);
            int num_fires; infile >> num_fires;
            vector<Point2D> fire_points;
            for (int i = 0; i < num_fires; ++i) {
                double x, y; infile >> x >> y;
                fire_points.push_back(Point2D(x, y));
            }
            int num_cameras; infile >> num_cameras;
            vector<Point2D> camera_positions;
            for (int i = 0; i < num_cameras; ++i) {
                double x, y; infile >> x >> y;
                camera_positions.push_back(Point2D(x, y));
            }
            infile.close();

            ifstream params(argv[3]);
            double radius, alpha; params >> radius >> alpha;
            params.close();

            vector<Vector2D> dirs = optimize_camera_angles(fire_points, camera_positions, radius, alpha);
            for (const auto& d : dirs) sol << fixed << setprecision(1) << d.get_x() << " " << d.get_y() << " ";
            sol << "\n";
        }
    }
    catch (const std::logic_error& e) { sol << "NOT_IMPLEMENTED\n"; }
    catch (const std::runtime_error& e) { sol << e.what() << "\n"; }
    catch (...) { sol << "ERROR\n"; }

    sol.close();
    return EXIT_SUCCESS;
}

File: LastSemesterInfo/AYDA Laboratorio/AYDA_Examen_4_geometria_computacional_(A2026)/mac/python/template.py
"""
@file template.py
@brief Examen Práctico 4: Emergencia en La Hechicera (Geometría Computacional)
Facultad de Ingeniería - ULA, Mérida.

INSTRUCCIONES:
- Completa las funciones marcadas con "TODO".
- NO modifiques las firmas de las funciones existentes.
- NO modifiques el bloque de comunicación final (Boilerplate).
"""

import sys
from typing import List
from lib import Point2D, Vector2D, Polygon
import geom

# ========================================================================
# EJERCICIO 1: Línea de Contención
# ========================================================================
def build_secure_perimeter(fire_points: List[Point2D], stations: List[Point2D]) -> Polygon:
    """
    Construye la frontera óptima alrededor del fuego y verifica que las
    instalaciones de la ULA estén a salvo.

    @throws RuntimeError("LA_HECHICERA_COMPROMETIDA") si una estación está en peligro.
    * HINT: Recuerda que la frontera más eficiente para encerrar un conjunto de puntos
      sin concavidades redundantes tiene un nombre específico en geometría.
    """
    # TODO: Implementar la construcción del perímetro y la validación de seguridad.
    raise NotImplementedError("NOT_IMPLEMENTED")

# ========================================================================
# EJERCICIO 2: Huella de Destrucción por Viento
# ========================================================================
def calculate_buffer_zone_area(p_fire: Polygon, wind_vector: Vector2D) -> float:
    """
    Calcula el área de la superficie total amenazada considerando el viento.

    * HINT: Un polígono no es más que una secuencia ordenada de vértices. Si conoces
      las coordenadas de los vértices del nuevo macro-polígono, existe un teorema
      algebraico para hallar su área exacta sin importar los polígonos que lo formen.
    """
    # TODO: Implementar el cálculo de la huella combinada y su área.
    raise NotImplementedError("NOT_IMPLEMENTED")

# ========================================================================
# EJERCICIO 3: Escuadrón de Drones Estáticos
# ========================================================================
def optimize_camera_angles(
    fire_points: List[Point2D],
    camera_positions: List[Point2D],
    radius: float,
    alpha: float
) -> List[Vector2D]:
    """
    Encuentra la combinación de orientaciones de drones que maximiza
    la vigilancia de focos únicos de incendio.

    * HINT: Dado que el número de drones es pequeño (N <= 8) y el espacio de
      decisiones es discreto (4 opciones por dron), evalúa cómo estructurar
      tu algoritmo para explorar el árbol de posibilidades sin omitir ninguna.
    """
    # TODO: Implementar la búsqueda de la configuración óptima mediante exploración recursiva.
    raise NotImplementedError("NOT_IMPLEMENTED")

# ========================================================================
# I/O BOILERPLATE (NO MODIFICAR ESTA SECCIÓN)
# Canal de comunicación automatizado con el Grader.
# ========================================================================
if __name__ == "__main__":
    if len(sys.argv) < 5: sys.exit(1)
    exercise, graph_file, params_file, sol_file = sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4]

    with open(sol_file, "w") as sol:
        try:
            if exercise == "ex1":
                with open(graph_file, "r") as f: tokens = f.read().split()
                if not tokens: raise ValueError()
                idx = 0
                num_fires = int(tokens[idx]); idx += 1
                fire_points = [Point2D(float(tokens[idx+i*2]), float(tokens[idx+i*2+1])) for i in range(num_fires)]; idx += num_fires*2
                num_stations = int(tokens[idx]); idx += 1
                stations = [Point2D(float(tokens[idx+i*2]), float(tokens[idx+i*2+1])) for i in range(num_stations)]

                perimeter = build_secure_perimeter(fire_points, stations)
                perimeter.for_each_vertex(lambda v: sol.write(f"{v.get_x():.4f} {v.get_y():.4f} "))
                sol.write("\n")

            elif exercise == "ex2":
                with open(graph_file, "r") as f: tokens = f.read().split()
                p_fire = Polygon()
                for i in range(int(tokens[0])): p_fire.add_vertex(Point2D(float(tokens[1+i*2]), float(tokens[2+i*2])))
                with open(params_file, "r") as p: ptokens = p.read().split()

                area = calculate_buffer_zone_area(p_fire, Vector2D(float(ptokens[0]), float(ptokens[1])))
                sol.write(f"{area:.6f}\n")

            elif exercise == "ex3":
                with open(graph_file, "r") as f: tokens = f.read().split()
                idx = 0
                num_fires = int(tokens[idx]); idx += 1
                fire_points = [Point2D(float(tokens[idx+i*2]), float(tokens[idx+i*2+1])) for i in range(num_fires)]; idx += num_fires*2
                num_cameras = int(tokens[idx]); idx += 1
                cameras = [Point2D(float(tokens[idx+i*2]), float(tokens[idx+i*2+1])) for i in range(num_cameras)]
                with open(params_file, "r") as p: ptokens = p.read().split()

                dirs = optimize_camera_angles(fire_points, cameras, float(ptokens[0]), float(ptokens[1]))
                for d in dirs: sol.write(f"{d.get_x():.1f} {d.get_y():.1f} ")
                sol.write("\n")

        except NotImplementedError: sol.write("NOT_IMPLEMENTED\n")
        except RuntimeError as e: sol.write(f"{str(e)}\n")
        except Exception: sol.write("ERROR\n")